### PR TITLE
Feat setup event bridge

### LINKED ISSUE #60: As a User, I receive a websocket notification if someone else joins the room
Othmane will think about the visual/UI implications later

### PR DESCRIPTION
**Pull request by MollardMichael**
This PR present an example of how to use EventBridge in a serverless application.

This allows to create a function called broadcast that will listen to all event and send it back to all the watcher in the same room.

Doing this will decouple the business logic of reacting to event from the broadcasting to all the watchers.

The negative impact is that  the round trip to database + call to the api gateway cost us 50 to 100 ms.

Closes #60

### CODE REVIEW

File: backend/functions/broadcast.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    const params: DocumentClient.QueryInput = {
+        TableName: process.env.ROOM_TABLE,
+        KeyConditionExpression: 'username = :u',
+        ExpressionAttributeValues: {
+            ':u': 'anonymous',
+        },
+    };
+
+    try {
+        const data = await dynamoDb.query(params).promise();
```
Comment (gjgd) [inline backend/functions/broadcast.ts:33]: Question: What is the data here? I'm not sure i understand what is posted to the socket

Comment (MollardMichael) [reply backend/functions/broadcast.ts:33]: For now, it's not much of anything. The putQuery is done in the connect function.

File: backend/functions/broadcast.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        await client
+            .postToConnection({
+                ConnectionId: event.detail.requestContext.connectionId,
+                Data: `found: ${JSON.stringify(data)}`,
+            })
+            .promise();
```
Comment (Startouf) [inline backend/functions/broadcast.ts:39]: Here in practice, we'll have to do this for every `watcher.connectionId` of the room right ? Do we have a way t do this in parallel ?

File: backend/serverless.yml [hunk 3]
```diff
@@ -51,10 +77,10 @@ functions:
...
           route: $connect
           # Only if we want authenticated connections
-          authorizer:
-            name: auth
-            identitySource:
-              - 'route.request.header.Auth'
+          # authorizer:
+          #   name: auth
+          #   identitySource:
+          #     - 'route.request.header.Auth'
```
Comment (Startouf) [inline backend/serverless.yml:83]: problem with this ?

File: backend/serverless.yml [hunk 4]
```diff
@@ -70,19 +96,11 @@ functions:
           route: $disconnect
   read:
     handler: functions/read.main
-    iamRoleStatementsInherit: true
-    iamRoleStatements:
-      - Effect: Allow
-        Action:
-          - dynamodb:Query
-        Resource:
-          - 'Fn::GetAtt': [RoomTable, Arn]
```
Comment (Startouf) [inline backend/serverless.yml:79]: why remove this ? If someone disconnects, then we need to clean his `watcher` key/value in `room.watchers` no ?

File: backend/functions/broadcast.ts (hunk 2 above)
Comment (MollardMichael) [reply backend/functions/broadcast.ts:39]: Yes we can do it in // because it's a prmise

File: backend/serverless.yml (hunk 3 above)
Comment (MollardMichael) [reply backend/serverless.yml:83]: It's not useful and will cost lambda invocation quota

File: backend/serverless.yml (hunk 4 above)
Comment (MollardMichael) [reply backend/serverless.yml:79]: Yes we'll need the dynamodb:PutItm Action. too. For now it's not useful.

File: backend/functions/broadcast.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            apiVersion: '2018-11-29',
+            endpoint: `https://${event.detail.requestContext.domainName}/${event.detail.requestContext.stage}`,
+        });
+    }
+
+    const params: DocumentClient.QueryInput = {
+        TableName: process.env.ROOM_TABLE,
+        KeyConditionExpression: 'username = :u',
+        ExpressionAttributeValues: {
+            ':u': 'anonymous',
```
[the remark this answers is not in the document]
Comment (MollardMichael) [reply backend/functions/broadcast.ts:28]: Once we merge your PR, I'll modify the code to use the new DB Schema.

File: backend/Makefile [hunk 6]
```diff
@@ -29,6 +29,9 @@ build: ## Run webpack build process (Useful for testing)
 deploy: ## Deploy to your personnal stage
 	sls deploy --stage=$(STAGE)
 
+f-deploy: ## Deploy to your personnal stage
```
[the remark this answers is not in the document]
Comment (MollardMichael) [reply backend/Makefile:32]: It's esaier this way for autocompletion.

File: backend/functions/broadcast.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Startouf) [reply backend/functions/broadcast.ts:28]: PR merged

File: backend/serverless.yml (hunk 3 above)
Comment (Startouf) [reply backend/serverless.yml:83]: then delete it ?

File: backend/libs/event-utuls.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { IApplicationEventWrapper, EventType, IEvent } from './response';
```
Comment (Startouf) [inline backend/libs/event-utuls.ts:1]: file name utuls -> utils

File: backend/libs/response.ts [hunk 8]
```diff
@@ -28,6 +29,31 @@ export interface IEvent extends Omit<APIGatewayProxyEvent, 'requestContext'> {
     source?: string;
 }
 
+export enum EventType {
```
Comment (Startouf) [inline backend/libs/response.ts:32]:
```suggestion
export enum BroadcastEventType {
```

**Pull request merged**